migration: make speed improvements

staebler opened the pull request:
This PR covers the 3 areas that we discussed changing to improve the speed of migration.

1) When saving owner references, we should only look at resource types that are known to have owner references to Hive resources. The `hiveutil v1migration save-owner-refs` command now has a hard-coded list of resource types that will be searched for owner references. This list can be overriden by passing in a file containing the list of resource types. Additionally, there is a `hiveutil v1migration discover-owned` command that can be run to collect the resource types owned by Hive resources. This command can be run on the int, stage, and prod clusters to ascertain prior to migration day which resource types should be included.
2) We should make use of parallelization for long tasks. The processes of saving owner references, restoring owner references, deleting Hive resources, and recreating Hive resources have all been modified to use concurrent goroutines.
3) We cannot afford to spend time verifying that the resources that were saved to disk match the resources that are currently in the cluster. This check has been removed from the process of deleting Hive resources.

staebler commented on contrib/pkg/v1migration/saveownerrefs.go:
```diff
@@ -1,29 +1,50 @@
+	hivev1alpha1.SchemeGroupVersion.WithResource("syncidentityproviders"),
+	hivev1alpha1.SchemeGroupVersion.WithResource("syncsetinstances"),
+	hivev1alpha1.SchemeGroupVersion.WithResource("syncsets"),
+	corev1.SchemeGroupVersion.WithResource("secrets"),
```
That is correct that there is only a single query to the API server: There is not another call to the API server for each individual secret.

I don't think that it is worthwhile to break the calls down into individual namespaces. I expect that the proportion of secrets in cluster namespaces will be pretty high, so we will not save much by skipping the other namespaces. However, we will add a lot of calls to the API server if we look at each namespace individually. If we are concerned about memory usage, a better approach would be to limit the number of responses allowed in each call to the API server. I will work on getting that done.

dgoodwin replied:
I'm not sure that's necessary because this all sounds fine. We're not likely to run out of memory are we?

staebler replied:
Too late. It's already done.

dgoodwin replied:
LOL

twiest commented on contrib/pkg/v1migration/recreateobjects.go:
```diff
@@ -82,58 +91,63 @@ func (o *RecreateObjectsOptions) Run() error {
+		logger.Info("created object")
 	}
-	logger.Info("created object")
+	wg.Done()
```
Same comment as above for using `defer` at the beginning of this function.

twiest commented on contrib/pkg/v1migration/deleteobjects.go:
```diff
@@ -90,68 +93,54 @@ func (o *DeleteObjectsOptions) Run() error {
+		kind := objFromFile.GetKind()
+		namespace := objFromFile.GetNamespace()
+		name := objFromFile.GetName()
+		logger := logger.WithFields(log.Fields{
```
Since I took the time to look it up, I thought I'd share here that logrus "is protected by a mutex for concurrent writes."

From: https://github.com/sirupsen/logrus/blob/master/README.md#thread-safety

twiest commented on contrib/pkg/v1migration/restoreownerrefs.go:
```diff
@@ -111,7 +136,7 @@ func (o *RestoreOwnerRefsOptions) Run() error {
 		}
 	}
-	return nil
+	wg.Done()
```
Same `defer` comment.

twiest commented on contrib/pkg/v1migration/saveownerrefs.go:
```diff
@@ -65,142 +89,149 @@ func (o *SaveOwnerRefsOptions) Validate(cmd *cobra.Command) error {
+		}
 	}
-	return refs
+	wg.Done()
```
Same `defer` comment.

twiest commented on contrib/pkg/v1migration/saveownerrefs.go:
```diff
@@ -65,142 +89,149 @@ func (o *SaveOwnerRefsOptions) Validate(cmd *cobra.Command) error {
 	}
-	return false
+	writer.Flush()
+	wg.Done()
```
Same `defer` comment.

twiest commented on hack/v1migration/README.md:
```diff
@@ -4,8 +4,10 @@
+  1. Run `./hack/v1migration/save.sh [workdir]` to backup all data to disk.
   1. Run `./hack/v1migration/remove_hiveadmission.sh` to remove the Hive validation webhooks.
-  1. Run `./hack/v1migration/delete.sh [workdir]` to backup all data to disk and remove Hive resources (CRs, CRDs, and admission webhooks).
+  1. Run `./hack/v1migration/delete.sh [workdir]` to remove Hive resources (CRs and CRDs).
```
Agreed. This is a much safer approach IMO.

twiest commented on contrib/pkg/v1migration/recreateobjects.go:
```diff
@@ -82,58 +91,63 @@ func (o *RecreateObjectsOptions) Run() error {
+		objChan <- &objFromFile
 	}
+	close(objChan)
+	recreateWG.Done()
```
Shouldn't this be `recreateWG.Wait()` ?

staebler replied:
Indeed it should be. Good catch.

staebler commented on contrib/pkg/v1migration/deleteobjects.go:
```diff
@@ -165,8 +154,9 @@ func (o *DeleteObjectsOptions) deleteObject(client dynamic.Interface, objFromFil
 		}
 		logger.Info("deleted object")
 	}
+	wg.Done()
```
Yes, I agree that this should use a defer. I know that I thought about that but I cannot remember why I left it the way it is. My best guess is that I was thinking that I did not want to mark it as done if the function exited early for some reason. However, causing a deadlock does not seem like a good resolution.

staebler commented on contrib/pkg/v1migration/deleteobjects.go:
```diff
@@ -90,68 +93,54 @@ func (o *DeleteObjectsOptions) Run() error {
+		}
+		gvr := hivev1alpha1.SchemeGroupVersion.WithResource(resourceForHiveKind(kind))
+		var resourceClient dynamic.ResourceInterface
+		if namespace != "" {
```
I guess I don't adhere to that best practice, but I do get where you are coming from. In this case, it is a matter of how you read it. I read it as `if the resource is namespaced then do this, else do that`, which is not a double negative.

Would you prefer that I change it to the following?

```
if len(namespace) > 0 {
 ...
} else {
 ...
}

twiest replied:
I was just pointing it out in case you agreed. I'm fine with leaving it as is since you don't agree. No change needed here.

staebler commented:
cf49c7b5...ca65e9bf
* Changed `waitGroup.Done` calls to be done in defers.
* Fixed incorrect call to `waitGroup.Done` with correct call to `waitGroup.Wait`.

twiest commented:
/lgtm
